### PR TITLE
Snapshot support

### PR DESCRIPTION
**Pull request by hallgren**

### CODE REVIEW

File: lib/sandthorn.rb [hunk 1]
```diff
@@ -39,12 +41,20 @@ def all aggregate_type
...
-    def find aggregate_id, aggregate_type
-      event_store_for(aggregate_type).find(aggregate_id, aggregate_type)
+    def find aggregate_id, aggregate_type, after_aggregate_version = 0
+      event_store_for(aggregate_type).find(aggregate_id, aggregate_type, after_aggregate_version)
     end
 
-    def save_snapshot(aggregate)
-      raise "Not Implemented"
+    def save_snapshot aggregate
+      if aggregate.aggregate_events.count > 0
```
Comment (jesjos) [inline lib/sandthorn.rb]: Suggestion: extract method: `if aggregate.unsaved_events?`

```ruby
def unsaved_events?
  aggregate_events.any?
end
```

File: lib/sandthorn/aggregate_root_base.rb [hunk 2]
```diff
@@ -83,13 +93,15 @@ def find id
...
+            current_aggregate_version = aggregate_from_snapshot.nil? ? 0 : aggregate_from_snapshot.aggregate_current_event_version
+            events = Sandthorn.find(aggregate_id, self, current_aggregate_version)
+            if aggregate_from_snapshot.nil? && events.empty?
               raise Errors::AggregateNotFound
             end
 
-            return aggregate_build events
-          rescue Exception
+            return aggregate_build events, aggregate_from_snapshot
+          rescue Exception => error
```
Comment (jesjos) [inline lib/sandthorn/aggregate_root_base.rb]: Why assign this if it isn't used?

File: lib/sandthorn/aggregate_root_base.rb [hunk 3]
```diff
@@ -83,13 +93,15 @@ def find id
 
         def aggregate_find aggregate_id
           begin
-            events = Sandthorn.find(aggregate_id, self)
-            unless events && !events.empty?
+            aggregate_from_snapshot = Sandthorn.find_snapshot(aggregate_id) if self.snapshot
+            current_aggregate_version = aggregate_from_snapshot.nil? ? 0 : aggregate_from_snapshot.aggregate_current_event_version
+            events = Sandthorn.find(aggregate_id, self, current_aggregate_version)
+            if aggregate_from_snapshot.nil? && events.empty?
```
Comment (jesjos) [inline lib/sandthorn/aggregate_root_base.rb]: You can write more confident code by wrapping `events` in an array:

```ruby
events = Array(Sandthorn.find(aggregate_id, self, current_aggregate_version)
raise Errors::AggregateNotFound unless events.any?
```

File: lib/sandthorn/snapshot_store.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    def save key, value
+      @store[key] = value
+    end
+
+    def find key
+      @store[key]
+    end
+  end
+end
```
Comment (jesjos) [inline lib/sandthorn/snapshot_store.rb]: Missing empty line

File: lib/sandthorn/aggregate_root_base.rb (hunk 3 above)
Comment (jesjos) [reply lib/sandthorn/aggregate_root_base.rb]: I’m sorry, sloppy reading. Nothing to see here!

Comment (hallgren) [reply lib/sandthorn/aggregate_root_base.rb]: No worries and great thanks for your insights in the language of Ruby.

**Pull request merged**